### PR TITLE
[Parse] Hoist diagnostics for named 'subscript'/'init'

### PR DESCRIPTION
**Pull request by rintaro**
`init something<T>(arg: T)` is probably more common mistake than `init<T> something(arg: T)`.

Also, removed `enum_element_decl_without_paren` diagnostics since it's unreachable.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Parse/ParsePattern.cpp [hunk 1]
```diff
@@ -612,27 +612,18 @@ Parser::parseSingleParameterClause(ParameterContextKind paramContext,
...
     // If we don't have the leading '(', complain.
     Diag<> diagID;
-    bool skipIdentifier = false;
     switch (paramContext) {
     case ParameterContextKind::Function:
     case ParameterContextKind::Operator:
       diagID = diag::func_decl_without_paren;
       break;
-    case ParameterContextKind::EnumElement:
-      diagID = diag::enum_element_decl_without_paren;
```
Comment (rintaro) [inline lib/Parse/ParsePattern.cpp]: @CodaFi Could you verify this is unreachable?

File: test/Parse/recovery.swift [hunk 2]
```diff
@@ -624,6 +624,8 @@ struct InitializerWithName {
 
 struct InitializerWithNameAndParam {
   init a(b: Int) {} // expected-error {{initializers cannot have a name}} {{8-9=}}
+  init y?(_ y: Int) {} // expected-error {{initializers cannot have a name}} {{8-9=}}
+  init z<T>(y: T) {} // expected-error {{initializers cannot have a name}} {{8-9=}}
```
Comment (jrose-apple) [inline test/Parse/recovery.swift]: I think I agree about `init z<T>`, but `init? y` seems more likely than `init y?` to me. Not that it really matters, this is all recovery.

File: lib/Parse/ParsePattern.cpp (hunk 1 above)
Comment (CodaFi) [reply lib/Parse/ParsePattern.cpp]: Confirmed.  Good catch.

File: test/Parse/recovery.swift (hunk 2 above)
Comment (rintaro) [reply test/Parse/recovery.swift]: Agreed.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**